src/object.rs

```diff
@@ -2163,27 +2124,52 @@ impl ObjectFactories {
         template: &mut [CK_ATTRIBUTE],
     ) -> Result<()> {
         let sensitive = obj.is_sensitive() || !obj.is_extractable();
-        let mut result = match self
-            .get_object_factory(obj)?
-            .check_get_attributes(template, sensitive)
-        {
-            Ok(()) => CKR_OK,
-            Err(e) => e.rv(),
-        };
+        let mut result = CKR_OK;
 
+        let factory_attrs =
+            self.get_object_factory(obj)?.get_data().get_attributes();
         let obj_attrs = obj.get_attributes();
+
         for ck_attr in template.iter_mut() {
-            match obj_attrs.iter().find(|a| a.get_type() == ck_attr.type_) {
+            let valid: bool;
+
+            /* check if this attribute is valid/allowed */
+            match factory_attrs.iter().find(|a| a.get_type() == ck_attr.type_) {
+                Some(fa) => {
+                    if sensitive && fa.is(OAFlags::Sensitive) {
+                        ck_attr.ulValueLen = CK_UNAVAILABLE_INFORMATION;
+                        if result == CKR_OK {
+                            result = CKR_ATTRIBUTE_SENSITIVE;
+                        }
+                        valid = false;
+                    } else {
+                        valid = true;
+                    }
+                }
                 None => {
-                    /* This attribute is not available on given object */
+                    /* This attribute is not valid for given object type */
                     ck_attr.ulValueLen = CK_UNAVAILABLE_INFORMATION;
-                    continue;
+                    if result == CKR_OK {
+                        result = CKR_ATTRIBUTE_TYPE_INVALID;
+                    }
+                    valid = false;
                 }
-                Some(attr) => {
-                    let attr_val = attr.get_value();
+            }
+
+            if !valid {
+                continue;
+            }
```

Comment:
Maybe worth adding some comment like

> The attribute is not available for export (either invalid or sensitive). Stop here!

The code is mostly easy to understand, but adding hints why exactly it is, will make it obvious on the first sight.

Reply:
ack, adding a blurb.